DTLOR_DP.py

```diff
@@ -107,313 +126,221 @@ def DP(hostTree, parasiteTree, phi, locus_map, D, T, L, Origin, R):
-                if co_min == A[(ep, eh, l_bottom)]:
-                    eventsDict[(vp, vh, l_bottom)].extend(Amin)
-               
-                #TODO: what does this do?
-                # do not allow top of gene tree handle to be an actual synteny
-                if Minimums[(vp, vh, l_bottom)] == Infinity or (ep=="pTop" and not top_is_star):
-                    del Minimums[(vp, vh, l_bottom)]
-                    del eventsDict[(vp, vh, l_bottom)]
-                # Compute O(ep, eh, l_top, l_bottom)
-                # Compute oBest[(vp, vh, l_top, l_bottom)], the source of O(ep, eh, l_top, l_bottom)
-                if vhIsATip: 
-                    O[(ep, eh, l_bottom)] = C[(ep, eh, l_bottom)]  
-                    oBest[(vp, vh, l_bottom)] = [(vp, vh, l_bottom)]              
+                C[(ep, eh, l_bottom)] = \
+                        combine_costs([A[(ep, eh, l_bottom)], duplications, transfers])
+
+                # Compute O[(ep, eh, l_bottom)]
+                # O is mapping_node -> (cost, [mapping_node])
+                if vh_is_a_tip: 
+                    O[(ep, eh, l_bottom)] = (C[(ep, eh, l_bottom)][0], [(vp, vh, l_bottom)])
                 else: 
-                
-                    #finds Minimum Cost for O
-                    O_list= [C[(ep, eh, l_bottom)], O[(ep, eh1, l_bottom)], O[(ep, eh2, l_bottom)]]
-                    O_min=min(O_list)
-                    O[(ep, eh, top_is_star, l_bottom)] = O_min     
-                    #finds the minimum switch locations for O
-                    oMin = O_list.index(O_min)
-                    if oMin == 0:
-                        oBest[(vp,vh, top_is_star, l_bottom)].append((vp, vh, top_is_star, l_bottom))
-                    if oMin == 1:
-                        oBest[(vp,vh, top_is_star, l_bottom)].extend(oBest[(vp, eh1, top_is_star, l_bottom)])
-                    if oMin == 2:
-                        oBest[(vp,vh, top_is_star, l_bottom)].extend(oBest[(vp, eh2, top_is_star, l_bottom)])
-            # Compute bestSwitch values
-            bestSwitch[(ep, "hTop", l_bottom)] = Infinity
-            bestSwitchLocations[(vp, hostTree["hTop"][1], l_bottom)] = [(None,None, None, None)]
+                    o_c = (C[(ep, eh, l_bottom)][0], [(vp, vh, l_bottom)])
+                    o_l = (O[(ep, eh1, l_bottom)])
+                    o_r = (O[(ep, eh2, l_bottom)])
+                    O[(ep, eh, l_bottom)] = combine_costs([o_c, o_l, o_r])
+
+            # Compute best_switch values
+            best_switch[(ep, "hTop", l_bottom)] = (Infinity, [])
             for eh in preorder(hostTree, "hTop"):
                 _, vh, eh1, eh2 = hostTree[eh]
-
-                # is vh a tip?
-                if eh1 == None: # then eh2 == None too and vh is a tip!
-                    vhIsATip = True
+                # Is vh a tip?
+                if check_tip(vh, eh1, eh2):
```

Comment:
since we are doing preorder here on species tree, we are probably checking whether vh is root, not tip~